### PR TITLE
ORC-469:[C++] Add EncodedStringVectorBatch to expose string dictionary in VectorBatch

### PR DESCRIPTION
**Pull request by yuruiz**

### CODE REVIEW

File: c++/include/orc/Vector.hh [hunk 1]
```diff
@@ -51,6 +51,8 @@ namespace orc {
     // whether there are any null values
     bool hasNulls;
 
+    bool isEncoded;
```
Comment (wgtmac) [inline c++/include/orc/Vector.hh]: add some comments to briefly describe what it means.

File: c++/include/orc/Vector.hh [hunk 2]
```diff
@@ -113,6 +115,36 @@ namespace orc {
...
 
+  struct StringDictionary {
+    StringDictionary(MemoryPool& pool);
+    DataBuffer<char> dictionaryBlob;
+
+    // Offset for each dictionary key entry.
+    DataBuffer<int64_t> dictionaryOffset;
+
+    void getValueByIndex(int64_t index, char*& valPtr, int64_t& length)
+    {
```
Comment (wgtmac) [inline c++/include/orc/Vector.hh]: move the curly brace to the the previous line to follow the coding style. same applies to code below

File: c++/src/ColumnWriter.cc [hunk 3]
```diff
@@ -840,15 +840,15 @@ namespace orc {
   /**
    * Implementation of increasing sorted string dictionary
    */
-  class StringDictionary {
+  class StringDictionaryImpl {
```
Comment (wgtmac) [inline c++/src/ColumnWriter.cc]: Rename it to SortedStringDictionary?

File: c++/include/orc/Reader.hh [hunk 4]
```diff
@@ -180,6 +180,16 @@ namespace orc {
      */
     RowReaderOptions& forcedScaleOnHive11Decimal(int32_t forcedScale);
 
+    /**
+     * Set enable encoding block mode.
+     */
+    RowReaderOptions& setEnableLazyDecoding(bool enable);
```
Comment (xndai) [inline c++/include/orc/Reader.hh]: If this is for dictionary encoding, can we be more specific here? For example, setEnableLazyDictionaryDecoding?

File: c++/include/orc/Type.hh [hunk 5]
```diff
@@ -64,7 +64,8 @@ namespace orc {
      * Create a row batch for this type.
      */
     virtual ORC_UNIQUE_PTR<ColumnVectorBatch> createRowBatch(uint64_t size,
-                                                             MemoryPool& pool
+                                                             MemoryPool& pool,
+                                                             bool encoded = false
```
Comment (xndai) [inline c++/include/orc/Type.hh]: rename into "dictionaryEncoded"?

File: c++/include/orc/Vector.hh [hunk 6]
```diff
@@ -113,6 +115,35 @@ namespace orc {
     DataBuffer<int64_t> length;
   };
 
+  struct StringDictionary {
+    StringDictionary(MemoryPool& pool);
+    DataBuffer<char> dictionaryBlob;
+
+    // Offset for each dictionary key entry.
+    DataBuffer<int64_t> dictionaryOffset;
```
Comment (xndai) [inline c++/include/orc/Vector.hh]: uint32_t? I assume 4G of dictionary size is more than enough.

File: c++/include/orc/Vector.hh [hunk 7]
```diff
@@ -113,6 +115,35 @@ namespace orc {
...
+
+    void getValueByIndex(int64_t index, char*& valPtr, int64_t& length) {
```
Comment (xndai) [inline c++/include/orc/Vector.hh]: index as uint32_t?

File: c++/src/ColumnReader.cc [hunk 8]
```diff
@@ -582,6 +579,20 @@ namespace orc {
...
+                                                  uint64_t numValues,
+                                                  char* notNull) {
+    ColumnReader::next(rowBatch, numValues, notNull);
+    notNull = rowBatch.hasNulls ? rowBatch.notNull.data() : nullptr;
+    rowBatch.isEncoded = true;
+
+    EncodedStringVectorBatch& batch = dynamic_cast<EncodedStringVectorBatch&>(rowBatch);
+    batch.dictionary = this->dictionary;
+
+    // Length buffer is reused to save dictionary entry ids
```
Comment (xndai) [inline c++/src/ColumnReader.cc]: You are not reusing, but create a new index vector, don't you?

File: c++/src/ColumnWriter.cc [hunk 9]
```diff
@@ -840,15 +840,15 @@ namespace orc {
   /**
    * Implementation of increasing sorted string dictionary
    */
-  class StringDictionary {
+  class SortedStringDictionary {
```
Comment (xndai) [inline c++/src/ColumnWriter.cc]: This renaming doesn't seem to relate to the purpose of this change. Suggest separate it.

File: c++/test/TestColumnReader.cc [hunk 10]
```diff
@@ -434,6 +434,89 @@ TEST(TestColumnReader, testDictionaryWithNulls) {
   }
 }
 
+TEST(TestColumnReader, testEncodedVectorWithNulls) {
```
Comment (xndai) [inline c++/test/TestColumnReader.cc]: Add test to cover case when dictionary encoding is not used, but lazy decoding option is set. Also cover composite types, such as list, map, etc.

File: c++/include/orc/Reader.hh [hunk 11]
```diff
@@ -180,6 +180,16 @@ namespace orc {
      */
     RowReaderOptions& forcedScaleOnHive11Decimal(int32_t forcedScale);
 
+    /**
+     * Set enable encoding block mode.
```
Comment (xndai) [inline c++/include/orc/Reader.hh]: Add more comments to explain what is "encoding block mode".

File: c++/include/orc/Vector.hh [hunk 12]
```diff
@@ -113,6 +115,35 @@ namespace orc {
...
+      }
+
+      int64_t* offsetPtr = dictionaryOffset.data();
+
+      valPtr = dictionaryBlob.data() + offsetPtr[index];
+      length = offsetPtr[index + 1] - offsetPtr[index];
+    }
+  };
+
+  struct EncodedStringVectorBatch : public ColumnVectorBatch {
```
Comment (xndai) [inline c++/include/orc/Vector.hh]: Add comment to explain the layout of EncodedStringVectorBatch.

File: c++/include/orc/Reader.hh (hunk 4 above)
Comment (yuruiz) [reply c++/include/orc/Reader.hh]: this is not only meant for dictionary decoding. This flag controls weather we use ColumnReader next() or nextEncoded() so I think we should leave it as is.

File: c++/include/orc/Type.hh (hunk 5 above)
Comment (yuruiz) [reply c++/include/orc/Type.hh]: Same as above. I think we better leave it as encoded for future support like RLE lazy decoding.

File: c++/include/orc/Vector.hh (hunk 6 above)
Comment (yuruiz) [reply c++/include/orc/Vector.hh]: Right now the RleDecoder only support int64_t*

File: c++/include/orc/Vector.hh (hunk 7 above)
Comment (yuruiz) [reply c++/include/orc/Vector.hh]: same as above, currently the RleDecoder only support int64_t, so we make it int64_t here as well.

File: c++/src/ColumnReader.cc (hunk 8 above)
Comment (yuruiz) [reply c++/src/ColumnReader.cc]: yes

File: c++/src/ColumnWriter.cc (hunk 9 above)
Comment (yuruiz) [reply c++/src/ColumnWriter.cc]: This change is to avoid the StringDictionary name conflicts cause we already declared StringDictionary in Vector.hh

**Pull request merged**